File: src/main/java/dev/jolvera/finjav/viewModels/BaseViewModel.java
package dev.jolvera.finjav.viewModels;

import javafx.application.Platform;
import javafx.beans.property.BooleanProperty;
import javafx.beans.property.SimpleBooleanProperty;
import javafx.beans.property.SimpleStringProperty;
import javafx.beans.property.StringProperty;

import java.util.function.Consumer;
import java.util.function.Supplier;

public abstract class BaseViewModel {
    protected final BooleanProperty isLoading = new SimpleBooleanProperty(false);
    protected final StringProperty errorMessage = new SimpleStringProperty("");

    public BooleanProperty isLoadingProperty() { return isLoading; }
    public StringProperty errorMessageProperty() { return errorMessage; }

    protected <T> void executeAsync(Supplier<T> supplier, Consumer<T> uiUpdate) {
        isLoading.set(true);
        errorMessage.set("");

        Thread.ofVirtual().start(() -> {
           try {
                T result = supplier.get();
                Platform.runLater(() -> {
                    uiUpdate.accept(result);
                    isLoading.set(false);
                });
           } catch (Exception e) {
                Platform.runLater(() -> {
                   errorMessage.set(e.getMessage());
                   isLoading.set(false);
                });
           }
        });
    }
}

File: src/main/java/dev/jolvera/finjav/viewModels/BaseViewModel.kt
package dev.jolvera.finjav.viewModels

import javafx.beans.property.BooleanProperty
import javafx.beans.property.SimpleBooleanProperty
import javafx.beans.property.SimpleStringProperty
import javafx.beans.property.StringProperty
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.cancel
import kotlinx.coroutines.javafx.JavaFx
import kotlinx.coroutines.launch
import kotlinx.coroutines.withContext

abstract class BaseViewModel {
    protected val isLoading = SimpleBooleanProperty(false)
    protected val errorMessage = SimpleStringProperty("")
    protected val scope = CoroutineScope(Dispatchers.JavaFx + SupervisorJob())

    fun IsLoadingProperty(): BooleanProperty = isLoading
    fun IsErrorProperty(): StringProperty = errorMessage

    protected fun <T> executeAsync(
        backgroundWork: suspend () -> T,
        uiUpdate: (T) -> Unit = {}
    ) {
        scope.launch {
            isLoading.value = true
            errorMessage.value = ""

            try {
                val result = withContext(Dispatchers.IO) {
                    backgroundWork()
                }
                uiUpdate(result)
            } catch (e: Exception) {
                errorMessage.value = e.message ?: "Unknown error"
            } finally {
                isLoading.value = false
            }
        }
    }

    open fun close() {
        scope.cancel()
    }
}